[CodeCompletion] Use substGenericArgs in getTypeOfMember

hamishknight opened the pull request:
For a GenericFunctionType, use `substGenericArgs` instead of `subst`, as the latter would form a bad generic signature if there were UnresolvedTypes present in the base type, causing the GSB to blow up when attempting to canonicalize it.

rdar://80635105

hamishknight commented:
@swift-ci please test

slavapestov reviewed (approved):
How close are we to banning subst() from being called on GenericFunctionType altogether?

hamishknight commented:
@slavapestov From a quick run of the test suite with an assertion in place, I *think* we're getting fairly close, `synthesizeDesignatedInitOverride` seems to be one of the main remaining offenders

slavapestov commented:
@hamishknight That's great!

hamishknight committed:
[CodeCompletion] Use substGenericArgs in getTypeOfMember

For a GenericFunctionType, use `substGenericArgs`
instead of `subst`, as the latter would form a bad
generic signature if there were UnresolvedTypes
present in the base type, causing the GSB to blow
up when attempting to canonicalize it.

rdar://80635105

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please smoke test Linux
